### DISCUSSION TITLE
open file in the correct editor window if multiple windows are open

### DISCUSSION THREAD

**Discussion by vesper8** [spatie/ray#363] [Ideas]
I know this may be tricky but it would be appreciated in the long term.

I always have multile VSCode windows opened for multiple projects. Currently when clicking on a file link in Ray it opens in the last-used VSCode window.. no matter what that window's project is.

I don't know if this is even possible but I think it would be nice if you tried to guess the project root (very easy for Laravel projects at least and probably so for most projects) and then attempted to open the file in that project window.

I know if you try to open vscode from the command line and that project is already opened in a vscode editor then it will be smart enough not to open a 2nd window, so perhaps clicking on a file could first open up a vscode instance for that project, which will either open a new vscode instance for that project, or simply focus the already-open one, and THEN try to open the file which would result in it opening in the correct project window

**Comment by AdrianMrn** [comment]
Sadly I don't think we have much control over this. VSCode exposes an URL pattern that we can fill, but from there on it's handled by vsc. Opening the URL through other means (eg browser), causes the same issue

The pattern for vscode is `vscode://file/path:line`, so you end up with something like `vscode://file//Users/adriaan/code/ray-demo-app/app/Console/Commands/ModelCommand.php:50`.

If you figure out a way to circumvent this and make vscode open the correct window, let me know and I'll definitely fix it!

**Reply by vesper8** [reply]
Well.. I do have a way of circumventing this on OSX at least. It only works from a local environment however but this is where Ray will be used 99% of the time. I use this method in several projects to open local Applications from a UI.. for example I have my own 'project manager' UI that I run locally and I can open browsers, editors, basically I can open anything on my Mac from a UI. Here's how I do it. In laravel I have a controller that looks like this

```
$path = '/path/to/project/root'
        $command = base_path('bash/openInApp.sh') . " '{$path}' 2>&1";
        exec($command);
```

And then I have this `openInApp.sh` somewhere which contains

```
    app="Visual Studio Code"
    cmd="tell application \"${app}\" to open \"${path}\""
    osascript -e "${cmd}"
```

This will either open a new vscode window for that project, or focus the already opened window.

You could then run a 2nd command to open the file in question which would open it in the last-focussed window.

You may even be able to do it in a single command.

**Comment by jasonvarga** [comment]
What do you call this `vscode://` uri pattern? I'm trying to find relevant issues on their repo, but either I find too many results or not enough. I was going to ask them if something like this is possible, or ask to be made possible.

**Reply by freekmurze** [reply]
I think it's called "url protocol"
